### ISSUE TITLE
Layout and behavior of enums with uninhabited fields in some variants

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#443]
This is about enums like the following:
```rust
enum E1 { A, C(!) }
enum E2 { A, C(!), D(i32, !) }
```

The first interesting point is that `E1` actually will not get any space assigned for storing a discriminant, it has size 0. The algorithm that assigns discriminants entirely skips `C`. In Miri we had to make SetDiscriminant throw UB early if the requested variant to set is uninhabited, since otherwise we get ICEs later.

However, `E2` has size 8, even though there is only a single valid variant and that variant has size 0. I *think* we currently always provide storage for all fields of all variants, even if they are uninhabited. This is useful because in theory it lets us compile `E2::D(f(), panic!())` into something that does in-place initialization:
```
let val: E2;
val.D.0 = f();
panic!();
SetDiscriminant(val, D); // this would be UB but we don't get here
```
(AFAIK we currently don't actually do that, we introduce temporaries instead.)

For structs we have decided that we will always have storage for all fields, and this is unavoidable since safe code can partially initialize a struct. However, safe code cannot partially initialize an enum, so we *could* soundly decide that `E2` has size 0. We have to decide between smaller enums and in-place initialization for arbitrary enums. (We can of course have small enums and then have an analysis that uses in-place initialization where possible -- but in generic MIR, it might not be possible to tell whether the variant we are about to initialize is inhabited.)

**Comment by Nadrieril** [comment]
Is it guaranteed (and documented somewhere) that `ReadDiscriminant` cannot return the discriminant of an uninhabited variant (on pain of UB)? Is that true even of `repr(whatever)` enums?

**Comment by RalfJung** [comment]
That is the current behavior as implemented in Miri, but I don't think we guarantee it. `ReadDiscriminant` is anyway more of an implementation detail, it doesn't show up directly in the surface syntax.

**Comment by Nadrieril** [comment]
Do we not specify what data a pattern-match accesses?

My real question is: is there or will there ever be a UB-free execution that can reach a match arm that matches on an uninhabited variant. Such an execution could look like the following (I'm assuming this example is fine for e.g. `T=u8`, do tell me if I'm wrong):
```rust
#![feature(never_type)]

#[repr(u8)]
enum Enum<T> {
    A = 0,
    B = 1,
    C(u8, T) = 2,
}

fn main() {
    let mut x: Enum<!> = Enum::A;
    unsafe { (&raw mut x).cast::<u8>().write(2u8) };
    match x {
        Enum::A => println!("got A!"),
        Enum::B => println!("got B!"),
        Enum::C(..) => println!("got C!"), // is this ever reachable?
    }
}
```

**Comment by RalfJung** [comment]
> Do we not specify what data a pattern-match accesses?

We specify basically nothing here. It's all details of how exactly `match` gets lowered to MIR / MiniRust, and that isn't really specified in any way.

Personally I think it's reasonable to say that GetDiscriminant will never return the discriminant of an uninhabited enum. Though this is yet another case of special-casing uninhabited types in the semantics (similar to https://github.com/rust-lang/unsafe-code-guidelines/issues/413 in that sense), which not everyone is happy with. And for `repr(C)` enums it is particularly tricky since we make so many layout guarantees about them.

**Comment by GKFX** [comment]
There has been a good bit of discussion of how this affects code which manually constructs an enum at [rust-lang/rust#120141](https://github.com/rust-lang/rust/issues/120141). (Also [an IRLO thread](https://internals.rust-lang.org/t/missing-layout-optimization-for-types-containing-infallible/22058).) Essentially, there would be a need for a fallible variant of `offset_of`, since if the enum is too small to contain the fields of one of its variants there is no valid offset to give for those fields. This is not insurmountably difficult to handle but it is likely to end up generally awkward for users.

Personally, I think that a type smaller than the fields it contains is just too surprising/strange of an optimization, and I doubt that manual construction of an enum is the only thing it would affect.

Are there other ways to achieve the same performance benefits without changing the enum's size? E.g., since the Rust calling convention is private to the compiler, having functions that take or return `Result<T, Uninhabited>` actually (in the ABI) take/return `T`. More generically, this would be - if there is exactly one inhabited variant, pass that directly.

**Comment by zachs18** [comment]
> ```rs
> enum E2 { A, B(!), C(i32, !) }
> ```
> 
> However, E2 has size 8, even though there is only a single valid variant and that variant has size 0. I think we currently always provide storage for all fields of all variants, even if they are uninhabited. This is useful because in theory it lets us compile E2::C(f(), panic!()) into something that does in-place initialization:

It seems like it should be possible to make `E2` be 4 bytes by having `C` not even hold space for a tag, since there is no sound program that could ever read or write the tag, and the tag is not exposed as a "real" field. (This optimization would not apply to `repr(C/Int)` enums, of course, only `repr(Rust)`).

I was looking at `compiler/rustc_abi/src/layout.rs` when working on https://github.com/rust-lang/rust/issues/135802 and realized that that seemed doable, so here's a branch with that change applied, and everything seems to work fine in ui tests: https://github.com/rust-lang/rust/compare/master...zachs18:rust:enum-uninhabited-variant-no-tag I haven't tested Miri with this change yet

**Comment by RalfJung** [comment]
> It seems like it should be possible to make E2 be 4 bytes while still having space for all fields by having C not hold space for a tag, since there is no sound program that could ever read or write the tag, and the tag is not exposed as a "real" field. (This optimization would not apply to repr(C/Int) enums, of course, only repr(Rust)).

Yeah that makes sense. We already do something like that in some cases -- AFAIK when computing a niched enum layout, variants with uninhabited fields are ignored when determining discriminants.

**Comment by Jules-Bertholet** [comment]
> However, safe code cannot partially initialize an enum

I think it would be nice if this became possible some day (e.g. with pattern types), which would rule out this optimization.

**Comment by chrysn** [comment]
Authoring libraries that are heavy in generics (whose performance in terms of memory use might depend on unused components to not matter, think zero cost abstractions), I'd prefer this to be resolved towards products of uninhabited types behaving like uninhabited types.

This would rule out the partial initialization in its simple form, but the compiler should have all the information to know if partial initialization can happen in this particular case -- whereas generic libraries would need to use quite cumbersome idioms (such as maybe-uninhibited traits with associated consts that are used as possibly 0-valued array lengths) to get the optimizations they expect.

**Comment by RalfJung** [comment]
> I'd prefer this to be resolved towards products of uninhabited types behaving like uninhabited types.

This thread is about sum types (enums). For product types (structs/tuples), this is already resolved and has been [documented in the Reference](https://doc.rust-lang.org/reference/type-layout.html#the-rust-representation) for a while: the fields of `repr(Rust)` structs do not overlap; even uninhabited fields have space reserved for them. Also see https://github.com/rust-lang/lang-team/pull/346.

**Comment by chrysn** [comment]
Thanks for the pointers; with this extra information, I think my concern still applies, but that is apparent only when I rephrase it to not bake in wrong assumptions:

Given that the sizes of product types (in explicit structs, tuples or as part of the set of fields inside an enum variant) can become large even when they are not inhabited, I think that it is important (in order to have simple-to-maintain generic code) that enums use knowledge of uninhabited variants to not grow due to their size.

In terms of the original example, I think that E2 should be optimized to just having the same storage properties of A.

(For as far as I am concerned, that is an optimization -- I don't need it guaranteed at all optimization levels, and in particular I don't need E2 to behave like `repr(transparent)` around A).

**Comment by CAD97** [comment]
> whereas generic libraries would need to use quite cumbersome idioms @chrysn 

FWIW, the cumbersome idioms apply in either direction. If partially uninhabited variants reserve space, then code wanting to get the smallest layout possible needs to do tricks. If partially uninhabited variants *don't* reserve space, then code wanting to do in-place initialization needs to do tricks.

The question becomes one of balancing two factors — which option is less surprising (thus will result in people writing UB less often) and which is easier to work around. As an additional factor, it is impossible for user code to currently perform piecewise initialization of an enum (no access to `SetDiscriminant`) without `MaybeUninit` (which is always inhabited).

- Reserving space for the inhabited fields is the "obvious" scheme, and thus won't incur surprise UB. Implementing space savings would most cleanly be expressed by fundamental autotrait specialization on inhabitedness.
- Eliminating partially uninhabited variants would make "obvious" in-place initialization potentially unsound. Working around this would mean checking inhabitedness (either by above autotrait specialization, `size_of` comparisons, or a new typeinfo function) and initializing all generics to the stack instead of in-place (to ensure the diverging initialization is called), which could in theory be itself unsound if a generic type relies on being in-place initialized in the expected context.

My current leaning is to be predictable and make layout reserve space for all inhabited fields and only reserve space for discriminants based on (fully) inhabited variants. But this is a loose leaning, and I could likely be convinced either way. Furthermore, the behavior when it's an enum variant field which is inhabited and when it's a struct field that's itself partially uninhabited don't *necessarily* need to be the same (even if it would be surprising for them to differ).

**Comment by CAD97** [comment]
Note: current behavior is a little bit inconsistent. `enum { A, B(u8, !) }` has size 2, reserving space for both the `u8` field and a `u8` discriminant, but `enum { A, B((), !) }` has size 0; i.e. we notice only when the partially uninhabited variant is zero sized that we don't need to reserve a discriminant for it.

This doesn't impact in-place partial initialization, but it does seem potentially relevant to the discussion. Ideally (at least imo) the discriminant elision should apply whenever a variant cannot be fully initialized, which would limit the cases where layout differs to those where any of the partially inhabited variants are either larger and/or more aligned than the enum without them.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/compiler-team/issues/922 is meant to fix that inconsistency.
